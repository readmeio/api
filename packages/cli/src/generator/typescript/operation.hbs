 module.
 [{{operationId}}]: function ()

 {
  operationId: 'list-data-sets',
  jsonSchema: null,
  operation: Operation {
    schema: {
      tags: [Array],
      operationId: 'list-data-sets',
      summary: 'List available data sets',
      responses: [Object]
    },
    api: {
      openapi: '3.0.1',
      servers: [Array],
      info: [Object],
      tags: [Array],
      paths: [Object],
      components: [Object]
    },
    path: '/',
    method: 'get',
    contentType: undefined,
    requestBodyExamples: undefined,
    responseExamples: undefined,
    callbackExamples: undefined
  }
}
