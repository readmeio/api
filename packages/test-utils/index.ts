export * from './fetch-mock.js';
export * from './load-spec.js';
