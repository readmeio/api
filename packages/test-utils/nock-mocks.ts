import type nock from 'nock';

import DatauriParser from 'datauri/parser';

export const response = { id: 123, name: 'Buster' };

export const responses = {
  all(this: nock.ReplyFnContext, uri: string, payload: nock.Body) {
    return [200, { uri, headers: this.req.headers, requestBody: payload }];
  },

  datauri(this: nock.ReplyFnContext, uri: string, payload: nock.Body) {
    const buffer = Buffer.from(payload as string, 'hex');
    const parser = new DatauriParser();

    return [
      200,
      {
        uri,
        requestBody: parser.format('.png', buffer).content,
        headers: this.req.headers,
      },
    ];
  },

  delay: (res: Record<string, unknown> | string, delay: number) => {
    return new Promise(resolve => {
      setTimeout(() => resolve(res), delay);
    });
  },

  headers: function responseHeaders(this: nock.ReplyFnContext) {
    return [200, this.req.headers];
  },

  multipart: function responseMultipart(this: nock.ReplyFnContext, uri: string, body: nock.Body): nock.ReplyFnResult {
    return [200, { uri, requestBody: body, headers: this.req.headers, boundary: body.split('\r\n')[0] }];
  },

  real: (payload: nock.Body) => {
    return () => [200, payload];
  },

  requestBody: (url: string, payload: nock.Body) => {
    return [200, { uri: url, requestBody: payload }];
  },

  searchParams: (uri: string) => {
    const res = new URL(`https://example.com${uri}`);
    return [200, `${res.pathname}${res.search}`];
  },

  url: (prop: keyof URL) => {
    return (uri: string) => {
      return [200, new URL(`https://example.com${uri}`)[prop]];
    };
  },
};
