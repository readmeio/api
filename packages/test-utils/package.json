{
  "private": true,
  "name": "@api/test-utils",
  "version": "7.0.0-alpha.0",
  "scripts": {
    "lint:types": "tsc --noEmit"
  },
  "dependencies": {
    "caseless": "^0.12.0",
    "datauri": "^4.1.0",
    "formdata-to-string": "^1.0.0",
    "vitest": "^0.34.5"
  },
  "devDependencies": {
    "@types/caseless": "^0.12.3",
    "oas": "^21.1.1",
    "typescript": "^5.2.2"
  },
  "prettier": "@readme/eslint-config/prettier"
}
