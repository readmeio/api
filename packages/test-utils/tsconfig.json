{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "lib": ["DOM", "ES2023"],
    "noEmit": true,
    "resolveJsonModule": true
  },
  "include": ["./**/*"]
}
