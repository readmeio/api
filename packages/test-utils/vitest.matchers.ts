import type { ParameterObject } from 'oas/rmoas.types';

import caseless from 'caseless';
import { expect } from 'vitest';

interface CustomMatchers<R = unknown> {
  /**
   * Assert that a given array within an OpenAPI definition has been dereferenced.
   */
  toBeDereferenced(): R;

  /**
   * Determine if a given `Headers` object has a given header matching a specific value.
   *
   * @example <caption>should match a value</caption>
   * expect(request.headers).to.have.header('connection', 'close');
   *
   * @example <caption>should match a regex</caption>
   * expect(response.headers).to.have.header('content-type', /application\/json(;\s?charset=utf-8)?/);
   *
   * @example <caption>should match one of many values</caption>
   * expect(request.headers).to.have.header('connection', ['close', 'keep-alive']);
   *
   */
  toHaveHeader(header: string, expected: RegExp | (string | number)[] | string): R;
}

declare module 'vitest' {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  interface Assertion<T = any> extends CustomMatchers<T> {}
  interface AsymmetricMatchersContaining extends CustomMatchers {}
}

expect.extend({
  toBeDereferenced(spec: ParameterObject[]) {
    const pass = !spec.filter(obj => '$ref' in obj).length;
    if (!pass) {
      return {
        message: () => 'expected the spec to be dereferenced',
        pass: false,
      };
    }

    return {
      message: () => 'expected the spec to not be dereferenced',
      pass: true,
    };
  },

  toHaveHeader(obj: Headers, header: string, expected: RegExp | (string | number)[] | string) {
    const headers = caseless(Object.fromEntries(Array.from(obj.entries())));

    // Header value should match a given regex.
    if (expected instanceof RegExp) {
      if (!expected.test(headers.get(header))) {
        return {
          message: () => `expected header to match ${expected.source}\n\nreceived: ${header}`,
          pass: false,
        };
      }

      return {
        message: () => `expected header to not match ${expected.source}\n\nreceived: ${header}`,
        pass: true,
      };
    }

    // Header value should exist in a given list.
    if (Array.isArray(expected)) {
      if (!expected.some(h => h === headers.get(header))) {
        return {
          message: () => `expected header to be one of the following: ${expected}\n\nreceived: ${header}`,
          pass: false,
        };
      }

      return {
        message: () => `expected header to not be one of the following: ${expected}\n\nreceived: ${header}`,
        pass: true,
      };
    }

    // Header value should match a given value.
    if (headers.get(header) !== expected) {
      return {
        message: () => `expected header to be ${expected}\n\nreceived: ${header}`,
        pass: false,
      };
    }

    return {
      message: () => `expected header not to be ${expected}\n\nreceived: ${header}`,
      pass: true,
    };
  },
});
