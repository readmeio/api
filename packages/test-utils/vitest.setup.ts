import nock from 'nock';
import { beforeAll, afterAll } from 'vitest';

beforeAll(() => {
  nock.disableNetConnect();
});

afterAll(() => {
  nock.cleanAll();
  nock.enableNetConnect();
});
