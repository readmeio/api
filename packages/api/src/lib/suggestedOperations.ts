import type { AuthForHAR } from '@readme/oas-to-har/lib/types';
import type Oas from 'oas';
import type Operation from 'oas/operation';

import APICore from '@readme/api-core';
import apiSnippetPlugin from 'httpsnippet-client-api';
import { Webhook } from 'oas/operation';

/**
 * Return a list of suggested and "good first issue" types of endpoints that we can build a post-SDK
 * code generation code snippet around to demo the power of `api`.
 *
 * The criteria for a suggested endpoint are the following:
 *  - GET request
 *  - Has no required query or body parameters
 *  - Endpoint has not been deprecated
 *
 * @param oas
 */
export function getSuggestedOperation(oas: Oas) {
  const suggested: Operation[] = [];

  Object.values(oas.getPaths()).forEach(path => {
    Object.values(path).forEach(operation => {
      // We don't really support webhooks with this library so if this operation is one we shouldn't
      // return it as a recommended operation.
      if (operation instanceof Webhook) {
        return;
      }

      if (operation.isDeprecated()) {
        return;
      }

      if (operation.method.toLowerCase() === 'get') {
        const hasRequiredParameters = operation.hasRequiredParameters();
        const hasRequiredRequestBody = operation.hasRequiredRequestBody();

        // If the criteria matches, push it into a recommended array
        if (!hasRequiredParameters && !hasRequiredRequestBody) {
          suggested.push(operation);
        }
      }
    });
  });

  if (suggested.length) {
    return suggested[0];
  }

  return false;
}

/**
 * Generate an example code snippet for a given (suggested) operation. We'll show this to users
 * post-codegeneration so they can see how to use the SDK we created for them.
 *
 * We're intentionally using `httpsnippet-client-api` and not `@readme/oas-to-snippet` here,
 * which would handle HAR and snippet generation, because don't need the entire `oas-to-snippet`
 * and `@readme/httpsnippet` libraries for what we're doing here.
 *
 * All we want to do is generate a very simple code example for `api` snippets and as we're
 * controlling which kinds of endpoints we're generating these for the HAR dataset we're working
 * with is a mostly fully known object.
 *
 */
export async function buildCodeSnippetForOperation(oas: Oas, operation: Operation, opts: { identifier: string }) {
  const core = new APICore();
  core.setSpec(oas);

  // If this endpoint has authentication on it then we should try to flesh out some placeholder
  // values in the `.auth()` SDK method for them so they can see how to use auth.
  let auth: AuthForHAR = {};
  const hasAuth = !!operation.getSecurity()?.length;
  if (hasAuth) {
    operation.getSecurityWithTypes().forEach(schemes => {
      if (!schemes) return;
      schemes.filter(Boolean).forEach(scheme => {
        if (!scheme) return;

        // eslint-disable-next-line no-underscore-dangle
        const schemeName = scheme.security._key;
        if (scheme?.type === 'Basic') {
          auth[schemeName] = { user: 'username', pass: 'password' };
        } else {
          auth[schemeName] = 'token';
        }
      });
    });

    auth = oas.getAuth(auth) as AuthForHAR;
  }

  const har = core.getHARForRequest(operation, {}, auth)?.log?.entries?.[0]?.request;
  if (!har) {
    return false;
  }

  const snippet = apiSnippetPlugin.client.convert(
    {
      ...har,
      cookiesObj: har.cookies.reduce((acc, { name, value }) => ({ ...acc, [name]: value }), {}),
      fullUrl: har.url,
      headersObj: har.headers.reduce((acc, { name, value }) => ({ ...acc, [name]: value }), {}),
      postData: {
        mimeType: 'application/json',
        text: '',
      },
      queryObj: har.queryString.reduce((acc, { name, value }) => ({ ...acc, [name]: value }), {}),
      url: har.url,

      // These aren't used in `httpsnippet-client-api` so we don't need to bother hooking them up.
      allHeaders: {} as never,
      uriObj: {} as never,
    },
    {
      api: {
        definition: oas.getDefinition(),
        identifier: opts.identifier,
        registryURI: opts.identifier,
      },
    },
  );

  return snippet;
}
