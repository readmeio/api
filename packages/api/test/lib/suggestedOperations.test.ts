import { loadSpec } from '@api/test-utils';
import Oas from 'oas';
import { Operation } from 'oas/operation';
import { describe, it, expect } from 'vitest';

import { getSuggestedOperation, buildCodeSnippetForOperation } from '../../src/lib/suggestedOperations.js';

describe('#getSuggestedOperation', () => {
  it('should retrieve a suggested operation', async () => {
    const spec = await loadSpec('@readme/oas-examples/3.0/json/readme.json').then(Oas.init);
    await spec.dereference();

    const suggested = getSuggestedOperation(spec);

    expect(suggested).toBeInstanceOf(Operation);
    expect((suggested as Operation).getOperationId()).toBe('getAPISpecification');
  });

  it('should not retrieve an operation from a spec thats only comprised of webhooks', async () => {
    const spec = await loadSpec('@readme/oas-examples/3.1/json/webhooks.json').then(Oas.init);
    await spec.dereference();

    const suggested = getSuggestedOperation(spec);
    expect(suggested).toBe(false);
  });
});

describe('#buildCodeSnippetForOperation', () => {
  it('should construct a code snippet for a given operation', async () => {
    const spec = await loadSpec('@readme/oas-examples/3.0/json/petstore.json').then(Oas.init);
    await spec.dereference();

    const operation = spec.operation('/pet/findByStatus', 'get');
    const snippet = await buildCodeSnippetForOperation(spec, operation, {
      identifier: 'petstore',
    });

    expect(snippet).toMatchSnapshot();
  });

  describe('should prefill in example auth auth tokens for endpoints that have auth', () => {
    it('basic auth', async () => {
      const spec = await loadSpec('@readme/oas-examples/3.0/json/readme.json').then(Oas.init);
      await spec.dereference();

      const operation = spec.operation('/api-specification', 'get');
      const snippet = await buildCodeSnippetForOperation(spec, operation, {
        identifier: 'readme',
      });

      expect(snippet).toContain(".auth('username', 'password')");
    });

    it('query auth', async () => {
      const spec = await loadSpec('@readme/oas-examples/3.0/json/security.json').then(Oas.init);
      await spec.dereference();

      const operation = spec.operation('/anything/apiKey', 'get');
      const snippet = await buildCodeSnippetForOperation(spec, operation, {
        identifier: 'security',
      });

      expect(snippet).toContain(".auth('token')");
    });
  });
});
