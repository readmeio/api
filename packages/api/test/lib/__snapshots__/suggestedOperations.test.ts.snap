// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#buildCodeSnippetForOperation > should construct a code snippet for a given operation 1`] = `
"import petstore from '@api/petstore';

petstore.auth('token');
petstore.findPetsByStatus()
  .then(({ data }) => console.log(data))
  .catch(err => console.error(err));"
`;
